### DISCUSSION TITLE
Use tmux pane/window for external terminal

### DISCUSSION THREAD

**Discussion by Abh15h3k** [mfussenegger/nvim-dap#656] [Q&A] [answered]
Currently i'm using alacritty as externalTerminal
```
dap.defaults.fallback.external_terminal = {
    command = '/opt/homebrew/bin/alacritty';
    args = {'-e'};
}
```

I use tmux on my main terminal (iterm2)
i was wondering if it was possible to use and split-pane as external terminal. i tried the following.
```
dap.defaults.fallback.external_terminal = {
    command = '/opt/homebrew/bin/tmux';
    args = {'split-pane'};
}
```

this works for most projects but for java projects with a big classpath it fails
i get `Terminal exited with status 1`

PS
the project which has the huge classpath is a project from work. so i can't share much details on it. if anyone has any suggestions please share

**Comment by Abh15h3k** [comment]
After a bit of tinkering a remembered that vscode put the class path to in an argfile and replaced `-cp ...` with `@<path-to-argfile>`
so i used a wrapper shell file with the following

```
tmux split-pane "$(python3 /tmp/argify.py "$@")"
```

argify.py
```
import sys

command = sys.argv[1:]
index = command.index("-cp")

f = open("/tmp/argfile", "w")
f.write(command[index])
f.write(" ")
f.write(command[index+1])
f.close()

del command[index]
command[index] = "@/tmp/argfile"

print(' '.join(command))
```

with this wrapper (or double wrapper?) the project runs.
any suggestions to make it cleaner ?

**Reply by mfussenegger** [reply]
java-debug supports a `shortenCommandLine` option which can be set to `jarmanifest` or `argfile`. That might help avoid your custom script.

**Reply by Abh15h3k** [reply]
> java-debug supports a `shortenCommandLine` option which can be set to `jarmanifest` or `argfile`. That might help avoid your custom script.

~how would i set the option?~
nevermind. it goes in the launch configuration.

Thanks for the help!

**Comment by Abh15h3k** [comment, marked as answer]
thanks to @mfussenegger

```
dap.configurations.configurations.java = {
    {
        ...
        shortenCommandLine = "argfile",
        ...
    }
}
```

same can be use while running test cases. just add the config while calling the api
`lua require'jdtls'.test_nearest_method({config = {shortenCommandLine="argfile"}})`

one caveat. if your test cases are path dependent (my work project loads some json files into mongodb using relative paths). you will have to enusre that the "terminal" starts at the correct root directory for tmux i do
```
dap.defaults.fallback.external_terminal = {
	command = "tmux",
	args = { "split-pane", "-c", "<desired-root-dir>"},
}
```

**Comment by Abh15h3k** [comment]
@mfussenegger is it possible to use the `shortenCommandLine` while running test cases via `lua require'jdtls'.test_nearest_method()` ? i get the same issue.

i used `dap.defaults.fallback.force_external_terminal = true` to force `externalTerminal`

**Reply by Abh15h3k** [reply]
figured it out. added to the answer

**Comment by AskaJohnny** [comment]
I have a problem when I use Telescope my dapui disappears, and then I can't connect to the original integrated terminal.

**Reply by AskaJohnny** [reply]
@mfussenegger  Big guy  I started the dap first and then used the telescope to locate the other file, and when I jumped to the other file, my original dapui disappeared, I opened it via dap.open, but dap-teminal couldn't connect to the previous one, including the dap ui scope also disappeared
